File: workspace/2023-04-01/numerical-sequence.prolog/factorial.pl

factorial(0, 1).

factorial(N, Fn) :-
    N > 0,
    is(M, N - 1),
    factorial(M, Fm),
    is(Fn, N * Fm).

factorials(N, Factorials) :-
    factorials(1, N, Factorials).

factorials(Min, Max, []) :-
    Min > Max.

factorials(Min, Max, [X|Xs]) :-
    factorial(Min, X),
    is(NextMin, Min + 1),
    factorials(NextMin, Max, Xs).

File: workspace/2024-04-27/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fn n -- fn )
    dup 0 = if
        drop
    else
        tuck * swap
        1 - _factorial
    endif
;

: factorial ( n -- fn )
    1 swap _factorial
;

: print_factorials_range recursive ( max min -- )
    over over < if
        drop drop
    else
        dup factorial .
        1 + print_factorials_range
    endif
;

: print_factorials ( n -- )
    s\" factorials:\n" type
    1 print_factorials_range
    s\" \n" type
;

